### ISSUE TITLE
Vertaisarviointi (viikko 4)

### ISSUE THREAD

**Issue by Veikkosuhonen** [attkauppi/MNIST_neuroverkko#1]
Projektisi on mielenkiintoinen ja olisi todella hienoa nähdä se lopullisessa muodossaan toiminnassa. Tämänhetkisessä rakenteessa ja suunnittelupäätöksissä on muutamia kohtia joista yritän antaa rakentavia neuvoja, joiden avulla projektin kehitys tästä edespäin saattaisi tulla helpommaksi. Nämä ovat tietenkin täysin omia näkemyksiäni. 

Semantiikkaa kannattaa lainata suosituista kirjastoista, kuten Kerasista. Siinä neuroverkkomalli luodaan esimerkiksi seuraavantapaisesti (pseudo):

```
layers = new Layer[] { lista kerros-objekteja. }
model = new NeuralNetwork(layers)
```

Näin käyttäjä saa rakentaa neuroverkon kerroksista ja tietää minkälainen jokainen kerros on, sen sijaan että antaa konstruktoriin lukuja syötteiden, kerrosten ja tulosteiden lukumäärän ja listan aktivaatiofunktioista. Se on hieman kömpelöä. Ja NeuronLayer-luokka sopii valmiiksi mainiosti tähän, `new NeuronLayer(3, new Sigmoid())` muistuttaa ihan Kerasin semantiikkaa `Dense(3,activation=”sigmoid”)`

Useimmissa kirjastoissa neuroverkkoa (tai muuta mallia) käytetään jotakuinkin näin:

```
model.fit(X, y)
output = model.predict(X)
```

Kutsuttaessa neuroverkkoa kannattaa käsitellä tilattomana funktiona. Sen sijaan että ensin input asetetaan, laskentametodi kutsutaan ja output haetaan, tämä voidaan tehdä luontevammin yhdellä metodilla. 

Poikkeustenkäsittelystä. Käytät try-catch -blokkeja poikkeusten nappaamiseen tilanteissa, joissa poikkeus aiheutuu siitä että ohjelma ei toimi niinkuin sen pitäisi. Tämä voi olla pidemmän päälle huono idea. Silloin kun jokin ohjelman osa toimii väärin, kannattaa pitää fail-fast periaate jotta perimmäinen ongelma voidaan korjata, sen sijaan että ongelma piilotetaan korjaavalla toiminnalla. Throwaa esimerkiksi tuo NeuronException silloin kun Neuronin inputiksi yritetään asettaa arvoa jonka indeksi on liian suuri, sen sijaan että kasvatat listaa. Tällöinhän `numberOfInputs` ja `inputs.size()` ovat muuttuneet eri kokoisiksi mutta ohjelma jatkaa toimintaansa.

Toisaalta välillä käytät try-catch-blokkeja tilanteissa, joissa mikään ei varsinaisesti mene pieleen, mutta vaihtoehtoista toimintaa tarvitaan. Tällöin käytä mielummin if-lauseita. Tiedät silloin paremmin (yleisesti ottaen) missä tilanteessa ohjelma tekee mitä, ja try-catch-blokkien rooli pysyy rajatunpana.

Vielä implementaatiosta, voisiko HiddenLayer, InputLayer ja Outputlayer muuttaa yhdeksi luokaksi? Kerroksen toiminnallisuushan on seuraava: sillä on tietty määrä syötteitä ja tulosteita ja se suorittaa jonkin operaation syötteillä ja tuottaa tulosteet. Kerrosta kutsutaan näin (esimerkiksi): `output = layer.process(input)`. Nyt kun neuroverkko koostuu listasta muutamia kerroksia, ei välttämättä tarvita erikseen Input ja Output tyyppisiä kerroksia. Verkolle annettu data yksinkertaisesti syötetään ensimmäiseen kerrokseen, tämän tuloste seuraavaan jne. kunnes viimeisen kerroksen tuloste annetaan verkon tulosteena. Kerrosten ei myöskään tarvitsisi tietää toisistaan, eli previousLayer ja nextLayer-kentät ja -metodit voisi poistaa mikä yksinkertaistaisi koodia. Neuroverkko itse voisi hoitaa kerrosten välisen datan siirron. 

Kiitos palautteen lukemisesta ja tsemppiä jatkoon.
